source/exe/process_wide.cc

```diff
@@ -9,16 +9,25 @@
 
 namespace Envoy {
 namespace {
+
+struct InitData {
+  uint32_t count_ ABSL_GUARDED_BY(mutex_){};
+  absl::Mutex mutex_;
+};
+
 // Static variable to count initialization pairs. For tests like
 // main_common_test, we need to count to avoid double initialization or
 // shutdown.
-std::atomic<uint32_t>& processWideInitialized() {
-  MUTABLE_CONSTRUCT_ON_FIRST_USE(std::atomic<uint32_t>);
-};
+InitData& processWideInitData() { MUTABLE_CONSTRUCT_ON_FIRST_USE(InitData); };
 } // namespace
 
 ProcessWide::ProcessWide() {
-  if (processWideInitialized()++ == 0) {
+  // Note that the following lock has the dual use of making sure that initialization is complete
+  // before a second caller can enter and leave this function.
+  auto& init_data = processWideInitData();
+  absl::MutexLock lock(&init_data.mutex_);
+
+  if (init_data.count_++ == 0) {
     ares_library_init(ARES_LIB_INIT_ALL);
     Event::Libevent::Global::initialize();
     Envoy::Server::validateProtoDescriptors();
```

Comment:
also this was pre-existing but I eel like areas should be initalized and destroyed, so that's fine as you have it, but there's no need to re-initialize Libevent and call validateProtoDescriptors, so maybe put another bool in InitData to capture whether they need to be called the second time around.

Reply:
OK I will add a TODO here as I'm trying to limit the changes here to fix an issue.